Add code review guideline

lukpueh opened the pull request:
Add guideline document about code reviews that may be used in addition to the development guideline document by code authors and code reviewers.

This document is largely based on a proposal by @aaaaalbert (see #17). It aims at making his proposal more succinct and adds post-review code author guidelines. Thanks, Albert!

I suggest @jhdalek55 and/or @JustinCappos as reviewers. Please review for conciseness and unnecessary redundancy with our [development guidelines](https://github.com/secure-systems-lab/lab-guidelines/blob/master/dev-workflow.md). Feel free to edit on this branch.

Commit:
Add code review guideline document

Add guideline document about code reviews that may be used in
addition to the development guideline document, by code authors
and code reviewers.

This document is largely based on a proposal  by @aaaaalbert
(see #17). It aims at making his proposal more succinct and adds
post-review code author guidelines.

Commit:
Link to code review guideline in dev workflow

lukpueh commented:
Note that the current version of the document does not include review examples, as suggested in #17. Although those might be helpful, I think the document is also valuable as is.

jhdalek55 commented:
I will try to take a look later today.

aaaaalbert reviewed (commented):
Thanks for digging this up (and attributing it 😄)!

Here's what I did to review your commit:
* Read over it
* Compare it with my old proposal in #17

aaaaalbert commented on code-review.md:
```diff
@@ -0,0 +1,64 @@
+The necessary prerequisite for a *line-by-line* review is that code authors
+make sure that the changes they propose actually make sense and work at a basic
+level. *Corollary: It should never happen that neither the author nor the
+reviewer actually tests the fix.* It is still okay to request feedback on work
```
I would rephrase this and explicitly (albeit redundantly) request that unit / integration tests must be run and must pass. --> remove "work at a basic level", add "the unit and integration tests pass".

Also, nobody knows what a corollary is 😉.

aaaaalbert commented on code-review.md:
```diff
@@ -0,0 +1,64 @@
+   means that it conforms to the [code
+   style](https://github.com/secure-systems-lab/code-style-guidelines) etc. *Do
+   not waste the reviewer's time with things you can easily get right in the
+   first place!*
```
Similar to the above, add "passing tests" as a requirement?

aaaaalbert commented on code-review.md:
```diff
@@ -0,0 +1,64 @@
+   attention:
+   > "Please do / look for / look at the following: ...."
+
+1. Feel free to suggest the review to someone, e.g. by using [GitHub's *review
```
Perhaps remove "Feel free to" so that a review is always suggest to a reviewer and doesn't end up in limbo.

aaaaalbert commented on code-review.md:
```diff
@@ -0,0 +1,64 @@
+
+1. Feel free to suggest the review to someone, e.g. by using [GitHub's *review
+   request*](https://help.github.com/en/articles/requesting-a-pull-request-review)
+   feature or by leaving a comment. Don't don't be afraid to remind them, if no
```
Typo: "Don't don't".

lukpueh replied:
But I did mean, that people should be afraid. :P (thanks for catching this)

aaaaalbert commented on code-review.md:
```diff
@@ -0,0 +1,64 @@
+
+__As code reviewer:__
+
+1. Inform the world about your review intent, e.g. by using [GitHub's
```
"Inform the world" feels quite informal. "State your intent"?

aaaaalbert commented on code-review.md:
```diff
@@ -0,0 +1,64 @@
+   review](https://help.github.com/en/articles/reviewing-proposed-changes-in-a-pull-request).
+   Be diligent, but also remember that the goal is not to demonstrate your
+   review skills but to integrate the proposed bug fix or feature. And be
+   respectful.
```
A great thing to say 👍

aaaaalbert commented on code-review.md:
```diff
@@ -0,0 +1,64 @@
+    - Reply to the comment, if it is unclear, or you think the reviewer got
+      something wrong.
+
+    - Push new commits that address the comment, otherwise. These commits
```
Permute: "Otherwise, push new commits [...]" so that it's clear that this is another conditional branch.

jhdalek55 commented:
Hey all. Sorry I have not checked in on this. We were scheduled to move last week and then didn't and that plus a bunch of other open commitments has slowed me down a bit. I will take a look either later today or tomorrow. Till then, it's nice seeing Albert's name pop up in my email. Hope all is well with you, Albert.

Commit:
Amplify responsibilities in code review guidelines

Following a review by @aaaaalbert this commit strengthens
several statements in code review guidelines (and fixes minor
mistakes).

lukpueh commented:
Thanks for the kind review, @aaaaalbert. I integrated your suggestions and tried to clarify a few things in 9e81129. Please, let me know what you think.

aaaaalbert commented:
My pleasure ❤️ , let's do this interactively!

jhdalek55 committed:
A few small tweaks 

I rewrote a handful of sentences for better flow and/or meaning.

JustinCappos committed:
Merge pull request #19 from jhdalek55/patch-1

A few small tweaks

lukpueh commented:
Thanks for the great word smithing, @jhdalek55, and for reviewing and merging, @JustinCappos! It seems that we are good to merge the entire document.
